src/box/lua/ctl.c

```diff
@@ -181,6 +188,7 @@ static const struct luaL_Reg lbox_ctl_lib[] = {
 	{"wait_rw", lbox_ctl_wait_rw},
 	{"on_shutdown", lbox_ctl_on_shutdown},
 	{"on_schema_init", lbox_ctl_on_schema_init},
+	{"on_schema_upgrade_end", lbox_ctl_on_schema_upgrade_end},
```

Comment:
I've several doubts regarding this trigger.

* It was not asked in the public API and maybe we should prefer to implement it internally (in some internal namespace and without doc request), at least for now.
* In fact, we don't ever need a new trigger, we can just set an `on_replace` trigger on the `_schema` system space and look for tuples with the `'version'` value in the first field (plus check `box.internal.schema_needs_upgrade`).
* It seems, it re-enters into the Lua VM from the FFI call. AFAIK, it is improper FFI usage and may cause unpredicatable behavior.
* It seems, the trigger is called on downgrade too.
* It seems, it is called on a failed upgrade too.

I propose to look toward `box.space._schema:on_replace`.

Reply:
Applied an approach with the `box.space._schema:on_replace` call.